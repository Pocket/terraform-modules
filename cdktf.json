{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
