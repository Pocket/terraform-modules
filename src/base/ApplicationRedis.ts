import {
  ApplicationElasticacheCluster,
  ApplicationElasticacheClusterProps,
  ApplicationElasticacheEngine,
} from './ApplicationElasticacheCluster';
import { Construct } from 'constructs';
import { DataAwsVpc } from '@cdktf/provider-aws/lib/data-aws-vpc';
import { ElasticacheReplicationGroup } from '@cdktf/provider-aws/lib/elasticache-replication-group';

const DEFAULT_CONFIG = {
  node: {
    size: 'cache.t3.micro',
    count: 2,
  },
};

export class ApplicationRedis extends ApplicationElasticacheCluster {
  public elasticacheReplicationGroup: ElasticacheReplicationGroup;

  constructor(
    scope: Construct,
    name: string,
    config: ApplicationElasticacheClusterProps
  ) {
    super(scope, name);

    // use default config, but update any user-provided values
    config = {
      ...DEFAULT_CONFIG,
      ...config,
    };

    const appVpc = ApplicationElasticacheCluster.getVpc(this, config);

    this.elasticacheReplicationGroup =
      ApplicationRedis.createElasticacheReplicationCluster(
        this,
        appVpc,
        config
      );
  }

  /**
   * Creates the elasticache cluster to be used
   * @param scope
   * @param appVpc
   * @param config
   * @private
   */
  private static createElasticacheReplicationCluster(
    scope: Construct,
    appVpc: DataAwsVpc,
    config: ApplicationElasticacheClusterProps
  ): ElasticacheReplicationGroup {
    const engine = ApplicationElasticacheEngine.REDIS;
    const port = ApplicationElasticacheCluster.getPortForEngine(engine);

    const { securityGroup, subnetGroup } =
      ApplicationRedis.createSecurityGroupAndSubnet(
        scope,
        config,
        appVpc,
        port
      );

    return new ElasticacheReplicationGroup(
      scope,
      'elasticache_replication_group',
      {
        replicationGroupId: `${config.prefix.toLowerCase()}`,
        replicationGroupDescription: `${config.prefix.toLowerCase()} | Managed by terraform`,
        nodeType: config.node.size,
        port: port,
        engineVersion:
          config.overrideEngineVersion === null ||
          config.overrideEngineVersion === undefined
            ? ApplicationRedis.getEngineVersionForEngine(engine)
            : config.overrideEngineVersion,
        parameterGroupName:
          config.overrideParameterGroupName === null ||
          config.overrideParameterGroupName === undefined
            ? ApplicationRedis.getParameterGroupForEngine(engine)
            : config.overrideParameterGroupName,
        automaticFailoverEnabled: true,
        subnetGroupName: subnetGroup.name,
        securityGroupIds: [securityGroup.id],
        tags: config.tags,
        applyImmediately: true,
        dependsOn: [subnetGroup, securityGroup],
        numberCacheClusters: config.node.count ?? 2,
        multiAzEnabled: true,
        provider: config.provider,
      }
    );
  }
}
