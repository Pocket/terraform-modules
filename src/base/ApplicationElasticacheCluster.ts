import { Resource } from 'cdktf';
import { VPC, ElastiCache } from '@cdktf/provider-aws';
import { Construct } from 'constructs';

export enum ApplicationElasticacheEngine {
  MEMCACHED = 'memcached',
  REDIS = 'redis',
}

export interface ApplicationElasticacheClusterProps {
  prefix: string;
  vpcId: string;
  subnetIds: string[];
  allowedIngressSecurityGroupIds: string[];
  node?: {
    /**
     * This is the size as defined here:https://aws.amazon.com/elasticache/pricing
     * Theres a lot here to make an enum for this..
     */
    size?: string;
    /**
     * Number of nodes that should exist in the cluster
     */
    count?: number;
  };
  tags?: { [key: string]: string };
}

/**
 * Generates an elasticache cluster with the desired engine
 */
export abstract class ApplicationElasticacheCluster extends Resource {
  protected constructor(scope: Construct, name: string) {
    super(scope, name);
  }

  /**
   * Gets a VPC
   * @param scope
   * @param config
   * @protected
   */
  protected static getVpc(
    scope: Construct,
    config: ApplicationElasticacheClusterProps
  ): VPC.DataAwsVpc {
    return new VPC.DataAwsVpc(scope, `vpc`, {
      filter: [
        {
          name: 'vpc-id',
          values: [config.vpcId],
        },
      ],
    });
  }

  /**
   * Returns the default port for the selected engine
   * @param engine
   * @private
   */
  protected static getPortForEngine(
    engine: ApplicationElasticacheEngine
  ): number {
    switch (engine) {
      case ApplicationElasticacheEngine.MEMCACHED:
        return 11211;
      case ApplicationElasticacheEngine.REDIS:
        return 6379;
    }
  }

  /**
   * Returns the default parameter group for the selected engine
   * @param engine
   * @private
   */
  protected static getParameterGroupForEngine(
    engine: ApplicationElasticacheEngine
  ): string {
    switch (engine) {
      case ApplicationElasticacheEngine.MEMCACHED:
        return 'default.memcached1.6';
      case ApplicationElasticacheEngine.REDIS:
        return 'default.redis6.x';
    }
  }

  /**
   * Returns the default engine version for the selected engine
   * @param engine
   * @private
   */
  protected static getEngineVersionForEngine(
    engine: ApplicationElasticacheEngine
  ): string {
    switch (engine) {
      case ApplicationElasticacheEngine.MEMCACHED:
        return '1.6.6';
      case ApplicationElasticacheEngine.REDIS:
        return '6.x';
    }
  }

  /**
   * Create a security group and a subnet group for Elasticache
   * @param scope
   * @param config
   * @param vpc
   * @param port
   * @protected
   */
  protected static createSecurityGroupAndSubnet(
    scope: Construct,
    config: ApplicationElasticacheClusterProps,
    vpc: VPC.DataAwsVpc,
    port: number
  ): {
    securityGroup: VPC.SecurityGroup;
    subnetGroup: ElastiCache.ElasticacheSubnetGroup;
  } {
    const securityGroup = new VPC.SecurityGroup(
      scope,
      'elasticache_security_group',
      {
        namePrefix: config.prefix,
        description: 'Managed by Terraform',
        vpcId: vpc.id,
        ingress: [
          {
            fromPort: port,
            toPort: port,
            protocol: 'tcp',

            //If we have a ingress security group it takes precedence
            securityGroups: config.allowedIngressSecurityGroupIds
              ? config.allowedIngressSecurityGroupIds
              : null,

            //IF we do not have a ingress security group lets all the whole vpc access
            cidrBlocks: config.allowedIngressSecurityGroupIds
              ? null
              : [vpc.cidrBlock],

            // the following are included due to a bug
            // https://github.com/hashicorp/terraform-cdk/issues/223
            description: null,
            ipv6CidrBlocks: null,
            prefixListIds: null,
          },
        ],
        tags: config.tags,
      }
    );
    const subnetGroup = new ElastiCache.ElasticacheSubnetGroup(
      scope,
      'elasticache_subnet_group',
      {
        name: `${config.prefix.toLowerCase()}-ElasticacheSubnetGroup`,
        subnetIds: config.subnetIds,
      }
    );
    return { securityGroup, subnetGroup };
  }
}
