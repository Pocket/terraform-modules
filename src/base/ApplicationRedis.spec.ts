import { Testing } from 'cdktf';
import { ApplicationElasticacheClusterProps } from './ApplicationElasticacheCluster';

import { ApplicationRedis } from './ApplicationRedis';

const BASE_CONFIG: ApplicationElasticacheClusterProps = {
  allowedIngressSecurityGroupIds: [],
  subnetIds: ['1234-123'],
  vpcId: 'cool-vpc',
  prefix: `abides-dev`,
};

describe('ApplicationRedis', () => {
  it('renders redis with minimal config', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRedis(stack, 'testRedis', BASE_CONFIG);
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders redis with tags', () => {
    const config: ApplicationElasticacheClusterProps = {
      ...BASE_CONFIG,
      tags: {
        letsgo: 'bowling',
        donnie: 'throwinrockstonight',
      },
    };
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRedis(stack, 'testRedis', config);
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders redis with node change', () => {
    const config: ApplicationElasticacheClusterProps = {
      ...BASE_CONFIG,
      node: {
        size: 'cache.m4.2xlarge',
        count: 5,
      },
    };
    const synthed = Testing.synthScope((stack) => {
      new ApplicationRedis(stack, 'testRedis', config);
    });
    expect(synthed).toMatchSnapshot();
  });
});
