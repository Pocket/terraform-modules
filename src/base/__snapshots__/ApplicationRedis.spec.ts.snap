// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationRedis renders redis with minimal config 1`] = `
"{
  \\"data\\": {
    \\"aws_vpc\\": {
      \\"testRedis_vpc_442CAD26\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"cool-vpc\\"
            ]
          }
        ]
      }
    }
  },
  \\"resource\\": {
    \\"aws_elasticache_replication_group\\": {
      \\"testRedis_elasticache_replication_group_E03AAD61\\": {
        \\"apply_immediately\\": true,
        \\"automatic_failover_enabled\\": true,
        \\"depends_on\\": [
          \\"aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7\\",
          \\"aws_security_group.testRedis_elasticache_security_group_36C6CB6D\\"
        ],
        \\"multi_az_enabled\\": true,
        \\"node_type\\": \\"cache.t3.micro\\",
        \\"number_cache_clusters\\": 2,
        \\"parameter_group_name\\": \\"default.redis6.x\\",
        \\"port\\": 6379,
        \\"replication_group_description\\": \\"abides-dev | Managed by terraform\\",
        \\"replication_group_id\\": \\"abides-dev\\",
        \\"security_group_ids\\": [
          \\"\${aws_security_group.testRedis_elasticache_security_group_36C6CB6D.id}\\"
        ],
        \\"subnet_group_name\\": \\"\${aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7.name}\\"
      }
    },
    \\"aws_elasticache_subnet_group\\": {
      \\"testRedis_elasticache_subnet_group_389F6AE7\\": {
        \\"name\\": \\"abides-dev-ElasticacheSubnetGroup\\",
        \\"subnet_ids\\": [
          \\"1234-123\\"
        ]
      }
    },
    \\"aws_security_group\\": {
      \\"testRedis_elasticache_security_group_36C6CB6D\\": {
        \\"description\\": \\"Managed by Terraform\\",
        \\"ingress\\": [
          {
            \\"cidr_blocks\\": null,
            \\"description\\": null,
            \\"from_port\\": 6379,
            \\"ipv6_cidr_blocks\\": null,
            \\"prefix_list_ids\\": null,
            \\"protocol\\": \\"tcp\\",
            \\"security_groups\\": [
            ],
            \\"self\\": null,
            \\"to_port\\": 6379
          }
        ],
        \\"name_prefix\\": \\"abides-dev\\",
        \\"vpc_id\\": \\"\${data.aws_vpc.testRedis_vpc_442CAD26.id}\\"
      }
    }
  }
}"
`;

exports[`ApplicationRedis renders redis with node change 1`] = `
"{
  \\"data\\": {
    \\"aws_vpc\\": {
      \\"testRedis_vpc_442CAD26\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"cool-vpc\\"
            ]
          }
        ]
      }
    }
  },
  \\"resource\\": {
    \\"aws_elasticache_replication_group\\": {
      \\"testRedis_elasticache_replication_group_E03AAD61\\": {
        \\"apply_immediately\\": true,
        \\"automatic_failover_enabled\\": true,
        \\"depends_on\\": [
          \\"aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7\\",
          \\"aws_security_group.testRedis_elasticache_security_group_36C6CB6D\\"
        ],
        \\"multi_az_enabled\\": true,
        \\"node_type\\": \\"cache.m4.2xlarge\\",
        \\"number_cache_clusters\\": 5,
        \\"parameter_group_name\\": \\"default.redis6.x\\",
        \\"port\\": 6379,
        \\"replication_group_description\\": \\"abides-dev | Managed by terraform\\",
        \\"replication_group_id\\": \\"abides-dev\\",
        \\"security_group_ids\\": [
          \\"\${aws_security_group.testRedis_elasticache_security_group_36C6CB6D.id}\\"
        ],
        \\"subnet_group_name\\": \\"\${aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7.name}\\"
      }
    },
    \\"aws_elasticache_subnet_group\\": {
      \\"testRedis_elasticache_subnet_group_389F6AE7\\": {
        \\"name\\": \\"abides-dev-ElasticacheSubnetGroup\\",
        \\"subnet_ids\\": [
          \\"1234-123\\"
        ]
      }
    },
    \\"aws_security_group\\": {
      \\"testRedis_elasticache_security_group_36C6CB6D\\": {
        \\"description\\": \\"Managed by Terraform\\",
        \\"ingress\\": [
          {
            \\"cidr_blocks\\": null,
            \\"description\\": null,
            \\"from_port\\": 6379,
            \\"ipv6_cidr_blocks\\": null,
            \\"prefix_list_ids\\": null,
            \\"protocol\\": \\"tcp\\",
            \\"security_groups\\": [
            ],
            \\"self\\": null,
            \\"to_port\\": 6379
          }
        ],
        \\"name_prefix\\": \\"abides-dev\\",
        \\"vpc_id\\": \\"\${data.aws_vpc.testRedis_vpc_442CAD26.id}\\"
      }
    }
  }
}"
`;

exports[`ApplicationRedis renders redis with tags 1`] = `
"{
  \\"data\\": {
    \\"aws_vpc\\": {
      \\"testRedis_vpc_442CAD26\\": {
        \\"filter\\": [
          {
            \\"name\\": \\"vpc-id\\",
            \\"values\\": [
              \\"cool-vpc\\"
            ]
          }
        ]
      }
    }
  },
  \\"resource\\": {
    \\"aws_elasticache_replication_group\\": {
      \\"testRedis_elasticache_replication_group_E03AAD61\\": {
        \\"apply_immediately\\": true,
        \\"automatic_failover_enabled\\": true,
        \\"depends_on\\": [
          \\"aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7\\",
          \\"aws_security_group.testRedis_elasticache_security_group_36C6CB6D\\"
        ],
        \\"multi_az_enabled\\": true,
        \\"node_type\\": \\"cache.t3.micro\\",
        \\"number_cache_clusters\\": 2,
        \\"parameter_group_name\\": \\"default.redis6.x\\",
        \\"port\\": 6379,
        \\"replication_group_description\\": \\"abides-dev | Managed by terraform\\",
        \\"replication_group_id\\": \\"abides-dev\\",
        \\"security_group_ids\\": [
          \\"\${aws_security_group.testRedis_elasticache_security_group_36C6CB6D.id}\\"
        ],
        \\"subnet_group_name\\": \\"\${aws_elasticache_subnet_group.testRedis_elasticache_subnet_group_389F6AE7.name}\\",
        \\"tags\\": {
          \\"donnie\\": \\"throwinrockstonight\\",
          \\"letsgo\\": \\"bowling\\"
        }
      }
    },
    \\"aws_elasticache_subnet_group\\": {
      \\"testRedis_elasticache_subnet_group_389F6AE7\\": {
        \\"name\\": \\"abides-dev-ElasticacheSubnetGroup\\",
        \\"subnet_ids\\": [
          \\"1234-123\\"
        ],
        \\"tags\\": {
          \\"donnie\\": \\"throwinrockstonight\\",
          \\"letsgo\\": \\"bowling\\"
        }
      }
    },
    \\"aws_security_group\\": {
      \\"testRedis_elasticache_security_group_36C6CB6D\\": {
        \\"description\\": \\"Managed by Terraform\\",
        \\"ingress\\": [
          {
            \\"cidr_blocks\\": null,
            \\"description\\": null,
            \\"from_port\\": 6379,
            \\"ipv6_cidr_blocks\\": null,
            \\"prefix_list_ids\\": null,
            \\"protocol\\": \\"tcp\\",
            \\"security_groups\\": [
            ],
            \\"self\\": null,
            \\"to_port\\": 6379
          }
        ],
        \\"name_prefix\\": \\"abides-dev\\",
        \\"tags\\": {
          \\"donnie\\": \\"throwinrockstonight\\",
          \\"letsgo\\": \\"bowling\\"
        },
        \\"vpc_id\\": \\"\${data.aws_vpc.testRedis_vpc_442CAD26.id}\\"
      }
    }
  }
}"
`;
